protocols/backend_spec.md

```diff
@@ -0,0 +1,1326 @@
+```
+The server's unwrapped response **MUST** match:
+```javascript
+exactly({
+    "supported": array(string),
+    "required": array(string)
+})
+```
+
+`"supported"` **SHOULD** be an array of specification names (for example "BE01" for this specification) for specifications on which the client may rely on the server conforming to. **NOTE** that the *client* is not required to conform to the listed specification.
+
+`"required"` **MUST** be an array of specification names (for example "BE01" for this specification) on which the client may rely on the server conforming but the server **will assume** that client the client conforms to.
+
+A specification **MUST** only be listed in at most one of these arrays. Specifications **SHOULD** only be listed in required if the specification indicates at the top "can be required by servers". Clients may refuse to operate with servers who do not support or require some set of specifications however they **SHOULD** only do this when the specification indicates "can be required by clients".
+
+Note that names are always two uppercase characters followed by two digits.
+
+# Basic Response
+Unless otherwise noted requests and responses are made in JSON, and the servers `Content-Type` header should indicate a mimetype of `application/json`. Requests and responses **SHOULD** be encoded in `UTF-8`.
+
+Unless otherwise noted the form of the JSON response **MUST** match either:
+```javascript
+{
+    "status": "error",
+    "error": string,
+    "error_description": optional(string),
+    "user_message": optional(string),
+    "error_data": optional(anything)
+}
+```
+in the case of an error. Where `error` gives a standard error name found from a specification and `error_description` some human readable error message. `user_message` gives some error message suitable for displaying to the end-user. 
+
+
+Or
+```javascript
+{
+    "status": "success",
+    "data": anything
+}
+```
+In the case of success. The specification may use the phrase "unwrapped response" or "unwrapped content" to refer to the contents of the `"data"` attribute (the response is required or assumed to be successful).
+
+
+When the specification mentions the server "returning error `x`" the response should match
+```javascript
+{
+    "status": "error",
+    "error": x,
+    "error_description": string
+}
+```
+
+A successful response without any data should match
+```javascript
+{
+    "status": "success",
+    "data": {}
+}
+```
+
+If a client request does not match a pattern that the specification requires that it matches the server **SHOULD** respond with error code 400 and error "invalid_request". eg, something like
+```javascript
+{
+    "status": "error",
+    "error": "invalid_request",
+    "error_description": "The client made an invalid request."
+}
+```
+## Internal errors
+At any time the server **MAY** respond to a request with response code 500 and return error "internal_server_error".
+
+# OAuth
```

Comment:
Alex already suggested in the slack that OAuth might be overkill. Seems to me like it is, but maybe we should vote on it?

Reply:
Yeah I agree that OAuth is a *HUGE* specification to bring in - In the spec I striped out as much of the OAuth stuff as possible (all stuff about client tokens is gone for instance), and only supported authentication using username and passwords (and refresh tokens) - I feel like this is pretty minimal and a a custom protocol would end up looking pretty similar. Would be interested in Alex's thoughts on this. But yeah a vote to get overall thoughts would be good.

Reply:
Alex had suggest JSON Web Tokens. I honestly don't know much about any of this, so probably better to just contact him directly.

Reply:
Hrm, In my understand JSON Web Tokens is a format of token and OAuth is way of getting tokens (without specifying what these look like) and in this way they are orthogonal.